v2/src/main/scala/coop/rchain/v2/casper/Casper.scala

```diff
@@ -0,0 +1,198 @@
+package coop.rchain.v2.casper
+import cats.syntax.all._
+import coop.rchain.v2.casper.data._
+import coop.rchain.v2.casper.syntax.all._
+
+/**
+ * @see [[https://github.com/cbc-casper/cbc-casper-paper/blob/master/cbc-casper-paper-draft.pdf Casper draft paper]]
+ *      for foundation of this implementation.
+ *
+ * @param dg            Dependency graph exposing justification structure.
+ * @param so            Safety oracle, defining fault tolerances.
+ * @param maxComplexity Maximum number of traversals through dependency graph.
+ * @tparam M            Type of a message.
+ * @tparam S            Type of a sender.
+ */
+final case class Casper[M, S](
+    dg: DependencyGraph[M, S],
+    so: SafetyOracle[M, S],
+    maxComplexity: Long = Long.MaxValue
```

Comment:
It is kind of wired to have a limit here.

Reply:
I'm trying to push a narrative that Casper is an algoritm, and we can limit its complexity if we want. So we cannot end up with Casper running for hours without knowing about that. We put a constraint on amount of resources that are allocated for Casper to search for a scope. If DAG is very diverged, Casper might not find a scope. Then we might increase complexity and try again, having the same DAG. Or just fail and throw some error.